api/basic/add-pin-router.ts

```diff
@@ -0,0 +1,250 @@
+    TABLE_NAME: table = '',
+    CLUSTER_BASIC_AUTH_TOKEN: token = '',
+    // set for testing
+    DYNAMO_DB_ENDPOINT: dbEndpoint = undefined,
+    INDEXER_ENDPOINT: indexerEndpoint = '',
+    PICKUP_ENDPOINT: pickupEndpoint = '',
+    BALANCER_RATE: balancerRate = 100
+  } = process.env
+
+  const authResponse = doAuth(event.headers.authorization)
+  if (authResponse != null) return authResponse
+
+  const dynamo = new DynamoDBClient({ endpoint: dbEndpoint })
+  const cid = event.pathParameters?.cid ?? ''
+  const origins = event.queryStringParameters?.origins?.split(',') ?? []
+
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (!table) {
+    return {
+      statusCode: 500,
+      body: JSON.stringify({ error: { reason: 'INTERNAL_SERVER_ERROR', details: 'TABLE must be set in ENV' } })
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (!cid) {
+    return { statusCode: 400, body: JSON.stringify({ error: { reason: 'BAD_REQUEST', details: 'CID not found in path' } }) }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (!indexerEndpoint) {
+    return { statusCode: 500, body: JSON.stringify({ error: { reason: 'INTERNAL_SERVER_ERROR', details: 'INDEXER_ENDPOINT not defined' } }) }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (!pickupEndpoint) {
+    return { statusCode: 500, body: JSON.stringify({ error: { reason: 'INTERNAL_SERVER_ERROR', details: 'PICKUP_ENDPOINT not defined' } }) }
+  }
+
+  if (!isCID(cid)) {
+    return {
+      statusCode: 400,
+      body: JSON.stringify({ error: { reason: 'BAD_REQUEST', details: `${cid} is not a valid CID` } })
+    }
+  }
+
+  for (const str of origins) {
+    if (!isMultiaddr(str)) {
+      return {
+        statusCode: 400,
+        body: JSON.stringify({ error: { reason: 'BAD_REQUEST', details: `${str} in origins is not a valid multiaddr` } })
+      }
+    }
+  }
+
+  try {
+    const res = await addPin({
+      cid, origins, dynamo, table, indexerEndpoint, pickupEndpoint, token, balancerRate: Number(balancerRate)
+    })
+    return { ...res, body: JSON.stringify(res.body) }
+  } catch (error) {
+    console.log(error)
+    return { statusCode: 500, body: JSON.stringify({ error: { reason: 'INTERNAL_SERVER_ERROR' } }) }
+  }
+}
+
+/**
+ * Handle a request to pin a CID.
+ * Returns existing Pin info if we have it, otherwise inserts a Pin record to
+ * DynamoDB and adds a message to SQS to kick off a pickup of the requested CID
+ * with optional source multiaddrs specified as origins list.
+ */
+export async function addPin ({
+  cid,
+  origins,
+  dynamo,
+  table,
+  indexerEndpoint,
+  pickupEndpoint,
+  token,
+  balancerRate
+}: AddPinInput): Promise<Response> {
+  const pinFromDynamo = await getPinFromDynamo(dynamo, table, cid)
+
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (!pinFromDynamo) {
```

Comment:
The flow is a little bit hard to follow, I'd use early returns (not necessary)

Reply:
cleaned up